railties/lib/rails/command.rb

```diff
@@ -37,8 +38,19 @@ def invoke(full_namespace, args = [], **config)
           command_name = namespace
         end
 
-        command_name, namespace = "help", "help" if command_name.blank? || HELP_MAPPINGS.include?(command_name)
-        command_name, namespace = "version", "version" if %w( -v --version ).include?(command_name)
+        if command_name.blank?
+          command_name, namespace = "help", "help"
+        elsif namespace == "application" && command_name == "application"
+          if args.empty? || (HELP_MAPPINGS & args).any?
+            command_name, namespace = "help", "help"
+          elsif (HELP_MAPPINGS & args).any?
+            command_name, namespace = "version", "version"
+          end
+        elsif HELP_MAPPINGS.include?(command_name)
+          command_name, namespace = "help_extended", "help"
+        elsif VERSION_MAPPINGS.include?(command_name)
+          command_name, namespace = "version", "version"
+        end
```

Comment:
This whole block is pretty ugly.
The reason is it that it has to handle multiple similar cases.
For example calling `rails -v` outside a Rails application, would call:
- Rails::Command.invoke ['-v']
- Rails::Command::VersionCommand.perform([])
- Rails::Command.invoke :application, ['--version']
- Rails::Command::ApplicationCommand.perform(['--version'])

I'd probably prefer to split this logic in to two methods (one inside a Rails app and one outside a Rails app)